openedx/openedx-platform
# [BB-3622] Add course creation condition for organisation

## Product Change Description


#### Current State

Studio, as of today doesn't have a way to restrict a user to create a course in a particular organization. What Studio provides right now is a `CourseCreator` permission which gives an `Admin` the power to grant a user the permission to create a course.

For example: If the  `Admin` has given a user `Spiderman` the permission to create courses, `Spiderman` can now create courses in any organization i.e `Marvel` as well as `DC`.

There is no way to restrict `Spiderman` from creating courses under `DC`.

#### Purpose of this PR

The changes done here gives `Admin` the ability to restrict a user on an Organization level from creating courses,

For example: Now, the `Admin` can give the user `Spiderman` the privilege of creating courses only under `Marvel` organization. The moment `Spiderman` tries to create a course under some other organization(i.e DC), we are showing the following error message:

![image](https://user-images.githubusercontent.com/7670449/128211264-ff9e3265-c462-4653-a687-c922c831e409.png)


###### Admin's Responsibility

With this PR the `Admin` can give users the permission to create courses under `any` organization or under a `specific` set of organizations. 

If `Admin` checks the `All Organization` checkbox then the `user` will be granted permission to create courses under any organization.

![Selection_024](https://user-images.githubusercontent.com/7670449/130602326-a2971e7a-610f-4138-a0d4-1595a9f592ee.png)

If `Admin` adds `specific` organizations under the selected lists then the `user` will only be able to create courses under those organizations.

![Selection_025](https://user-images.githubusercontent.com/7670449/130603248-9dc81093-b912-4c2d-9dec-c66de81355fd.png)


**There are few edge cases that can be encountered and below are the screenshot of how we inform the Admin/User about it**


**When `All Organization` is unchecked and no `Organization` is added to the list**

![image](https://user-images.githubusercontent.com/7670449/128209710-11b91724-e973-400c-9eb6-4488fbd25559.png)

**When `All Organization` is checked and specific `Organization` are added as well**

![image](https://user-images.githubusercontent.com/7670449/128210135-a13128ad-a60e-4763-b925-93db9e562956.png)

**Error message when the user enters organization name for which user doesn't have permission**

![image](https://user-images.githubusercontent.com/7670449/128211264-ff9e3265-c462-4653-a687-c922c831e409.png)




## Engineering Implementation

The condition added to ensure that if the feature is enabled
user will not be able to create the course outside of the organisation
in which they belong.

**JIRA tickets**: [BB-3622](https://tasks.opencraft.com/browse/BB-3622) 

~~**Discussions**: Link to any public dicussions about this PR or the design/architecture. Otherwise omit this.~~

**Dependencies**: None

**Screenshots**: 
![image](https://user-images.githubusercontent.com/7670449/108390150-61d3c100-7236-11eb-914c-efb1b8781425.png)



~~**Sandbox URL**: TBD - sandbox is being provisioned.~~

~~**Merge deadline**: "None" if there's no rush, "ASAP" if it's critical, or provide a specific date if there is one.~~

**Testing instructions**:

1. Get the master devstack up and running
2. Pull this branch in edx-platform
3. Make sure to run the migration
3. Drop in the studio shell `make dev.shell.studio`
4. vim /edx/etc/studio.yml and set ENABLE_CREATOR_GROUP = True
The below changes are also required to be done
```diff
diff --git a/cms/envs/devstack.py b/cms/envs/devstack.py
index 9d610b31f0..b548167a35 100644
--- a/cms/envs/devstack.py
+++ b/cms/envs/devstack.py
@@ -152,7 +152,7 @@ CREDENTIALS_SERVICE_USERNAME = 'credentials_worker'
 FEATURES['CERTIFICATES_HTML_VIEW'] = True
 
 ########################## AUTHOR PERMISSION #######################
-FEATURES['ENABLE_CREATOR_GROUP'] = False
+FEATURES['ENABLE_CREATOR_GROUP'] = True
 
 ################### FRONTEND APPLICATION PUBLISHER URL ###################
 FEATURES['FRONTEND_APP_PUBLISHER_URL'] = 'http://localhost:18400'
diff --git a/cms/envs/devstack_decentralized.py b/cms/envs/devstack_decentralized.py
index f7d61d41a1..f115c9e56a 100644
--- a/cms/envs/devstack_decentralized.py
+++ b/cms/envs/devstack_decentralized.py
@@ -107,7 +107,7 @@ CREDENTIALS_SERVICE_USERNAME = 'credentials_worker'
 FEATURES['CERTIFICATES_HTML_VIEW'] = True
 
 ########################## AUTHOR PERMISSION #######################
-FEATURES['ENABLE_CREATOR_GROUP'] = False
+FEATURES['ENABLE_CREATOR_GROUP'] = True
 
 ################### FRONTEND APPLICATION PUBLISHER URL ###################
 FEATURES['FRONTEND_APP_PUBLISHER_URL'] = 'http://localhost:18400'
```

5. ~~In the `FEATURES` dictionary add  `RESTRICT_NON_ORG_COURSE_CREATION: true`~~
6. Now create a user
7. ~~Go to the `admin` page and in localhost:18010/admin/student/courseaccessrole/ create a new  entry, add the user, give it instructor privilege, don't add the `Org` yet~~
8. Log in the studio and request for permission to create the course
9. Log in with a user who is a staff and superuser in another browser(edx on devstack)
10. Go to [http://localhost:18010/admin/course_creators/coursecreator/](http://localhost:18010/admin/course_creators/coursecreator/)
11. Here you can see the permission, all_organization and organization field
10. If you enable all_organzitaions field it will create a CourseCreator Role and works accordingly(The old flow)
11. If you disable all_organization field then you need to add Organizations, this will create OrganiztionContentCreator Role for that organization
12. Then you can try creating a course and in the Org field you need to enter org short name(Same as the one you gave permission to create in the previous role), this is case sensitive field and it will create the course
13. If any other Org is entered, it will not create the course and will through error
14. Another area that can be tested is on enabling `all_organization field`, course_content_group role s created in [http://localhost:18010/admin/student/courseaccessrole/](http://localhost:18010/admin/student/courseaccessrole/) and on disabling it and adding org, OrgContentCreator role is created.

Additional Manual Tests

1. In the course creator admin, change the status to granted and check if roles are created
2. Change the status to anything but granted and check if roles gets removed, roles can be checked in [http://localhost:18010/admin/student/courseaccessrole/](http://localhost:18010/admin/student/courseaccessrole/)

**Author notes and concerns**:

1.  It's a in discovery feature
2. The update to m2m field is non atomic operation, hence https://timonweb.com/django/many-to-many-field-save-method-and-the-django-admin-caveat/ was taken in consideration to obtain the desired result.

**Reviewers**
